### PR TITLE
OSS Delta Sharing Server: Adds api to accept cdf query

### PR DESCRIPTION
**Pull request by linzhou-db**
OSS Delta Sharing Server: Adds api to accept cdf query

- @Get("/shares/{share}/schemas/{schema}/tables/{table}/changes")
- Parse url parameters and construct the cdfoptions map
- Add classes DeltaErrors and DeltaDataSource for some exceptions and constants   
- Add `DeltaSharedTable.queryCDF` and return not implemented exception

### CODE REVIEW

**Comment by zhuansunxt** [comment]
Hi @linzhou-db , could you have a standalone PR on https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md with the new API's specification? It's better to agree on specification before adding its implementation on the server IMO.

**Comment by linzhou-db** [comment]
> Hi @linzhou-db , could you have a standalone PR on https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md with the new API's specification? It's better to agree on specification before adding its implementation on the server IMO.

sg, will do

**Comment by linzhou-db** [comment]
> > Hi @linzhou-db , could you have a standalone PR on https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md with the new API's specification? It's better to agree on specification before adding its implementation on the server IMO.
> 
> sg, will do

Sent out https://github.com/delta-io/delta-sharing/pull/137

File: server/src/main/scala/io/delta/standalone/internal/DeltaErrors.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.standalone.internal
+
```
Comment (zhuansunxt) [inline server/src/main/scala/io/delta/standalone/internal/DeltaErrors.scala:18]: Should we put those inside a `DeltaErrors` object and make those functions?

File: server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala [hunk 2]
```diff
@@ -241,6 +244,37 @@ class DeltaSharingService(serverConfig: ServerConfig) {
...
+      includeFiles = true,
+      getCdfOptionsMap(
+        Option(startingVersion),
+        Option(endingVersion),
+        Option(startingTimestamp),
+        Option(endingTimestamp)
+      )
+    )
+    logger.info(s"Took ${System.currentTimeMillis - start} ms to load the table cdf " +
+      s"and sign ${actions.length - 2} urls for table $share/$schema/$table")
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala:274]: yes, this is copied from L243

File: server/src/main/scala/io/delta/standalone/internal/DeltaErrors.scala (hunk 1 above)
Comment (linzhou-db) [reply server/src/main/scala/io/delta/standalone/internal/DeltaErrors.scala:18]: yes, will update.

File: server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala [hunk 3]
```diff
@@ -333,6 +368,72 @@ object DeltaSharingService {
...
+    }
+  }
+
+  private[server] def getCdfOptionsMap(
+    startingVersion: Option[String],
+    endingVersion: Option[String],
+    startingTimestamp: Option[String],
+    endingTimestamp: Option[String]): Map[String, String] = {
+    checkCDFOptionsValidity(startingVersion, endingVersion, startingTimestamp, endingTimestamp)
+    (if (startingVersion.isDefined) {
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala:410]: I'll try.

File: server/src/main/scala/io/delta/standalone/internal/DeltaErrors.scala (hunk 1 above)
Comment (linzhou-db) [reply server/src/main/scala/io/delta/standalone/internal/DeltaErrors.scala]: updated.

File: server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (linzhou-db) [reply server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala]: updated.

File: server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala [hunk 4]
```diff
@@ -343,6 +376,62 @@ object DeltaSharingService {
...
+    }
+    if (endingVersion.isDefined && endingTimestamp.isDefined) {
+      throw DeltaCDFErrors.multipleCDFBoundary("ending")
+    }
+    if (startingVersion.isEmpty && startingTimestamp.isEmpty) {
+      throw DeltaCDFErrors.noStartVersionForCDF
+    }
+    if (startingVersion.isDefined) {
+      try {
+        startingVersion.get.toLong
```
Comment (chakankardb) [inline server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala:396]: Should we validate timestamp?
If it gets validate elsewhere, perhaps add a comment.

File: server/src/main/scala/io/delta/standalone/internal/DeltaCDFErrors.scala [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+package io.delta.standalone.internal
+
+object DeltaCDFErrors {
+  def multipleCDFBoundary(position: String): Throwable = {
+    new IllegalArgumentException(s"Multiple $position arguments provided for CDF read. Please " +
+      s"provide one of either ${position}Timestamp or ${position}Version."
+    )
+  }
+
+  def noStartVersionForCDF: Throwable = {
```
Comment (chakankardb) [inline server/src/main/scala/io/delta/standalone/internal/DeltaCDFErrors.scala:26]: Should we call it noStartParamForCDF?

File: server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala (hunk 4 above)
Comment (linzhou-db) [reply server/src/main/scala/io/delta/sharing/server/DeltaSharingService.scala:396]: Added a comment, same as the code in our repo, it's validated in the cdc reader.

File: server/src/main/scala/io/delta/standalone/internal/DeltaCDFErrors.scala (hunk 5 above)
Comment (linzhou-db) [reply server/src/main/scala/io/delta/standalone/internal/DeltaCDFErrors.scala:26]: I more prefer to keep it the same as DeltaErrors in our code base.

Comment (chakankardb) [reply server/src/main/scala/io/delta/standalone/internal/DeltaCDFErrors.scala:26]: sg.

**Review by chakankardb — APPROVED** [review]

**Comment by linzhou-db** [comment]
@zhuansunxt you still want to take a look?

**Review by zhuansunxt — APPROVED** [review]
LGTM

**Comment by Basem-Gaber** [comment]
Why is the concept of table changes coupled with the CDF ?

Isn't it possible to extract the info of the difference between a version and another by only knowing which files were added and which were removed ?

The API can return this metadata which should be sufficient for the client to use to replicate the changes, correct ?

**Comment by linzhou-db** [comment]
Hi Basem, 
Thanks for your interest and question.


> Why is the concept of table changes coupled with the CDF ?

CDF is short for Change Data Feed, it IS the concept of table changes.


> Isn't it possible to extract the info of the difference between a version and another by only knowing which files were added and which were removed ?
> 
> The API can return this metadata which should be sufficient for the client to use to replicate the changes, correct ?

Technically yes, while CDF is a well developed, organized, and well maintained feature for this purpose, so good to just use it.

**Pull request merged**